동기화 메모장 [STEP 1] 차차, 예거

Jager-yoo opened the pull request:
안녕하세요! 코리! @corykim0829 
4기 캠퍼 차차와 예거입니다!

<동기화 메모장> 스텝 1 완료 후 PR 제출합니다! 😄 
앞으로 3주간 잘 부탁드리겠습니다~~ 👍🏻 

# Step 1 PR

## 🔥 고민했던 점

### 1️⃣ 다크모드 대응 

- App이 light mode와 dark mode에 대응할 수 있도록 구현하였습니다.
- 두 mode에서 각 UI 요소들의 시각적인 부분에서 불편함이 없는지 전반적으로 확인 후  색상을 선택하여 구현하였습니다. 

|light mode|dark mode|
|:---:|:---:|
|![](https://i.imgur.com/ACzf8p6.png)|![](https://i.imgur.com/pXxuGDa.png)|

<br>

### 2️⃣ 키보드가 화면을 가리지 않도록 구현

- `UITextView`를 터치하여 키보드가 올라올 때 일부 컨텐츠를 가리는 현상이 있었습니다.
- `UIResponder`의 `keyboardWillShowNotification`, `keyboardWillHideNotification` notification을 받아 키보드가 등장하고 사라질 때 실행될 메서드를 각각 구현했습니다. 
- 우선 키보드가 완전히 등장했을 때의 높이를 notification의 userInfo를 통해 얻었습니다. 이후 키보드의 높이 값을 `UITextView`의 `contentInset.bottom`에 할당하여 키보드의 높이만큼 `UITextView`의 inset을 조정해주어 보고자 하는 컨텐츠가 가려지는 문제를 해결했습니다. 

<p align="left"><img src="https://i.imgur.com/MSQX7Bt.gif" width="50%"></p>

<br>

### 3️⃣ App 을 처음 실행했을 때 보이는 화면

- App을 처음 실행했을 때, secondary view에 빈 화면이 보이는 것이 아니라 tableView의 가장 첫 번째 cell에 해당하는 데이터의 내용이 보이도록 구현했습니다.
- 빈 화면이나, 별도의 안내 문구보다는 첫 데이터의 내용을 보여주는 것이 자연스럽다고 판단하여 이와 같이 구현했습니다.

<br>

### 4️⃣ 멀티태스킹 대응

- Multitasking 을 하게 되면 `UISplitViewController`가 collapsed되어 `single container`가 되는데, 이 때 초기에 secondary view가 우선적으로 보이게 되는 현상이 있었습니다. 
- 이러한 현상이 문제는 아니지만, 메모의 내용이 보이는 것 보다 primary에 해당하는 메모 리스트인 tableView가 보이는 것이 조금 더 적합할 것 같다고 판단하여 `UISplitViewControllerDelegate`의 메서드를 활용하여 collapsed 되었을 때 primary view가 우선적으로 보이도록 구현해주었습니다.
- 또한 collapsed 유무를 확인하기 위해 `isCollapsed`를 사용하여 상황에 맞게 메서드를 실행할 수 있도록 구분해주었습니다. 

<p align="left"><img src="https://i.imgur.com/wha0IMw.gif" width="50%"></p>

```swift
extension NotesSplitViewController: UISplitViewControllerDelegate {
    func splitViewController(_ svc: UISplitViewController, topColumnForCollapsingToProposedTopColumn proposedTopColumn: UISplitViewController.Column) -> UISplitViewController.Column {
        return .primary
    }
}
```

<br>

### 5️⃣ TableView extension 구현

- UITableView를 구성하기 위해 cell의 identifier를 알고 있어야 한다는 점이 부담스러울 수 있다고 판단하여 이를 몰라도 cell을 사용할 수 있도록 UITableView extension을 구현하여 사용해주었습니다.
- 아래와 같이 구현하는 경우 두 가지 이점을 얻을 수 있는 것 같습니다. 
  - cell의 identifier를 신경쓰지 않아도 된다.
  - 반환되는 cell이 optional 타입이 아니다.

```swift
extension UITableView {
    func register<T: UITableViewCell>(cellWithClass name: T.Type) {
        register(T.self, forCellReuseIdentifier: String(describing: name))
    }
    
    func dequeueReusableCell<T: UITableViewCell>(withClass name: T.Type, for indexPath: IndexPath) -> T {
        guard let cell = dequeueReusableCell(withIdentifier: String(describing: name), for: indexPath) as? T else {
            fatalError("cell을 dequeue하는데 실패했습니다.")
        }
        
        return cell
    }
}
```

<br>

## 🙋🏻‍♂️ 질문하고 싶은 부분

### 1️⃣ 지역화 구현 방법

- `timeIntervalSince1970` Int 값을, 날짜로 알아볼 수 있는 String 으로 변환하는 과정에서 `locale`을 사용했습니다.
- 이때, `Locale.current` 프로퍼티를 활용하여, 현재 아이패드 기기의 `설정 언어`를 따라가는 `지역화`를 시도했습니다.
- `지역`의 경우, 한국이나 일본으로 바꾸면 current 에 정상적으로 지역 코드가 바뀌는 반면
- `iPad 언어` 설정은 한국어나 일본어로 바꿔봐도, current 에 찍히는 건 `영어(en)` 였습니다.
- 그래서 `날짜 지역화가 항상 영문으로 고정되는 이슈`가 있습니다.
- 🤔 질문 : 현업에서 `지역화`를 고민하실 때는 어떤 코드를 사용하시는지 궁금합니다!
사용자의 `기기 언어`를 기준으로 지역화를 하는 게, 지역을 기준으로 하는 것 보다 더 정확하다고 생각하는데요, 지역화와 관련해 공유해주실 수 있는 팁이 있으실까요~?

<p align="left"><img src="https://user-images.githubusercontent.com/71127966/153207506-cbdd9825-d2fb-47e2-943b-8ad65deaa857.png" width="80%"></p>
<p align="left"><img src="https://user-images.githubusercontent.com/71127966/153208694-1b08e02f-3c57-4e16-9af9-0edc8240ece1.png" width="80%"></p>
<p align="left"><img src="https://user-images.githubusercontent.com/71127966/153209564-b8f917aa-b6bd-429a-8971-cb2582de5749.png" width="40%"></p>

<br>

### 2️⃣ PrimaryVC 에서 선택된 cell 데이터를 SecondaryVC 로 보내는 방법

- PrimaryVC 는 테이블뷰입니다. cell 이 선택되면(didSelect), 해당 cell 에 해당하는 Memo 데이터를 SecondaryVC 의 UITextView 에 보낼 방법을 고민했습니다.
- 처음엔 스플릿뷰컨을 경유하지 않고, `delegate 프로퍼티`를 만들어서 PrimaryVC -> SecondaryVC 로 바로 데이터를 보내는 구조를 생각했는데요.
둘이 부모~자식 뷰컨이라면 모르겠지만, 스플릿뷰컨이라는 공통 부모를 가진 `동등한 레벨의 뷰컨`이라는 점 때문에 delegate 을 사용하는 게 맞는지 잘 모르겠더라구요! 😅 
- 저희는 결국 `메서드 주입` 방법을 택했는데요.
PrimaryVC 와 SecondaryVC 의 공통 부모인 `스플릿뷰컨`에 아래와 같이 SecondaryVC 로 데이터를 보내는 메서드를 만들고 테이블뷰의 `didSelectRowAt`메서드 에서 호출해줬습니다.
- 요약하자면 `PrimaryVC -> 스플릿뷰컨 -> SecondaryVC` 방향으로 선택된 cell 데이터를 전달하는 구조입니다.
- 🤔 질문 : `스플릿뷰컨` 계층에서 PrimaryVC ~ SecondaryVC 사이의 데이터 전달은 어떤 방식으로 하는 게 좋을까요?
저희가 선택한 `메서드 주입` 방식을 개선해볼 만한 포인트를 알려주실 수 있을까요? 🙏🏻

```swift
func showSecondaryView(with memo: Memo) {
    secondaryViewController.updateMemo(text: memo.body) // SecondaryVC 로 메서드 주입
    show(.secondary) // Single Container 모드에서 테이블뷰의 cell 이 선택됐을 때 화면 push 해주는 역할
}
```

<br>

### 3️⃣ Secondary View 의 width 가 조정될 때, UITextView 내부의 텍스트가 어색하게 움직이는 이슈

- 글로 설명드리는 것보다, 아래 gif 와 함께 설명드리는 게 편할 것 같습니다!
- 스플릿뷰컨에 내장되어 있는 `displayModeButton` 버튼을 눌러서, `사이드바`를 숨겼다가, 다시 펼칠 때 마다 우측에 위치한 텍스트들의 줄바꿈이 조정되는데요!
- width 가 넓어질 때는 버튼을 누르자마자 줄바꿈이 `즉시` 조정되서 자연스러운 반면
(마치 viewWillAppear 처럼 동작하는 느낌)
- width 가 좁아질 때는 버튼을 누르면, 텍스트 양옆이 잘려나가면서 -> 일단 width 만을 맞춘 뒤에, 애니메이션이 끝나고 텍스트 줄바꿈이 일어납니다.
(마치 viewDidAppear 처럼 동작하는 느낌)
- 저희는 줄바꿈과 관련해서 따로 설정을 하진 않았는데요, 애플이 기본적으로 구현해둔 `줄바꿈 처리`가 이렇게 미흡하진 않을 것 같다는 의심이 들었습니다. 😅 
- 🤔 질문 : 이런 어색한 줄바꿈 애니메이션은 어떻게 고쳐볼 수 있을까요?
조금 욕심을 내자면, 글자 요소 단위로 물흐르듯이 줄바꿈이 바뀌는 효과를 줘보고 싶은데... 관련하여 참고할 자료가 있을까요?

<p align="left"><img src="https://user-images.githubusercontent.com/71127966/153214885-1854ee84-a285-42ea-9e8b-9aee89387df5.gif" width="50%"></p>

ChaminLee committed:
feat: SwiftLint 적용, iOS deployment target 14.3 -> 14.0

ChaminLee committed:
file: gitignore Pods 주석 해제

Jager-yoo committed:
fix: Build Phases 의 Run Script 변경을 통한 SwiftLint 활성화

ChaminLee committed:
feat: 스토리보드 제거 및 UISplitViewController 생성

- Device를 iPad만 지원하도록 설정
- SceneDelegate내 window 설정
- ViewController 제거
- Deployment target 14.0 -> 14.3

Jager-yoo committed:
file: MVC 패턴대로 폴더 구조 변경

Jager-yoo committed:
feat: SplitView 에 Primary, Secondary View 세팅

- SceneDelegate 에서 rootViewController 설정할 때 SplitViewController 의 style 이니셜라이저 통해 doubleColumn 으로 생성
- SwiftLint 에서 trailing_whitespace 규칙 제외함
- Intel 맥북과 M1 맥북 간의 building 오류가 있어서, Excluded Architectures - Any iOS Simulator SDK -> 'arm64' 로 설정함

ChaminLee committed:
feat: sample.json 파일 추가

ChaminLee committed:
feat: Memo 타입 구현

Jager-yoo committed:
feat: JSONConverter 타입 구현

ChaminLee committed:
refactor: JSONConverter generics를 메서드가 아닌 타입에 적용

ChaminLee committed:
test: JSONConverter의 decode 메서드 테스트 코드 추가

Jager-yoo committed:
refactor: JSONConverter 타입 내에서 DateFormatter 역할 제거

Jager-yoo committed:
feat: MemoTableViewCell 타입 구현

ChaminLee committed:
feat: UITableView extension을 통한 register, dequeue 구현

- UITableViewCell의 identifier를 알고 있지 않아도 타입의 이름을 사용하여 register, dequeue되도록 구현
- UITableViewCell내 identifier 타입 프로퍼티 제거

Jager-yoo committed:
feat: UITableViewDataSource 메서드 구현

- [MemoTableViewCell 타입] Memo 타입을 파라미터로 받아서 셀의 Labels text 채워주는 메서드 구현
- 메서드 이름 변경 configureCell() -> configureCellLayout()
- [PrimaryTableViewController 타입] 샘플 JSON 데이터를 디코딩해서 프로퍼티로 저장해주는 메서드 구현하고 viewDidLoad() 에서 호출

ChaminLee committed:
feat: Int extension을 통한 dateString 연산 프로퍼티 구현

- 기기의 locale에 따라 날짜의 표현 방식이 다르게 보이도록 구현

Jager-yoo committed:
feat: Navigation Bar 의 아이템 세팅

- title 부여
- rightButtonItem 에 + 아이콘 추가
- [NotesSplitViewController] preferredSplitBehavior 를 .tile 로 설정

ChaminLee committed:
feat: SecondaryViewController UI 구현

- UITextView 구현
- rightNavigationItem 버튼 구현

Jager-yoo committed:
feat: 테이블뷰의 didSelectRowAt 메서드 구현

- 선택된 indexPath 에 위치한 메모 데이터를 이용해 SecondaryVC 의 textView 를 업데이트 하는 로직 구현
- textView 에 focus 가 들어간 상태로(커서 넣은 상태) text 를 업데이트하면, 커서가 textView 의 최하단으로 내려가는 이슈를 해결하기 위해, 업데이트 할 때 마다 resignFirstResponder() 메서드를 호출해서 focus를 제거함

ChaminLee committed:
feat: Keyboard show/hide에 따른 textView 높이 조정

- 키보드에 의해 콘텐츠가 가려지는 문제 해결

Jager-yoo committed:
feat: UITableViewCell 의 selectedBackgroundView 색상을 -> systemBlue 변경

- UITableView.Style -> insetGrouped 변경

ChaminLee committed:
file: AppIcon 추가

ChaminLee committed:
feat: TableViewCell의 selected 상태를 유지하기 위한 기능 추가

- TableView내에 최신 indexPath 저장을 위한 currentSelectedIndexPath 생성
- 위 기능을 하는 selectCell 메서드를 viewWillAppear에 삽입하여 primary view가 사라지더라도 selected 상태를 유지하도록 구현

Jager-yoo committed:
Revert "feat: TableViewCell의 selected 상태를 유지하기 위한 기능 추가"

This reverts commit 7577ef996ab5fdab59b57298dfe4ea1c562fff76.

Jager-yoo committed:
feat: 멀티 태스킹 모드가 돼서 화면이 Collapsed 될 때 PrimaryVC 가 보이도록 구현

- 최초 화면이 Regular Width 를 가졌다면, SecondaryVC 에 첫 번째 cell 의 text 가 보이도록 분기처리함

corykim0829 commented on CloudNotes/CloudNotes/Controller/NotesSplitViewController.swift:
```diff
@@ -0,0 +1,39 @@
+
+class NotesSplitViewController: UISplitViewController {
+    private var primaryTableViewController = PrimaryTableViewController(style: .insetGrouped)
+    private var secondaryViewController = SecondaryViewController()
```
Primary Secondary라고 하면, 이 클래스만 봤을 때에는 어떤 역할을 하는 코드인지 알기 어려울 것 같아요. 이름을 바꾸면 좀 더 명확한 코드가 될 것 같아요!

Jager-yoo replied:
코멘트 감사합니다!
지금 뷰컨들의 네이밍에는 `위치`는 명확하게 드러나지만 `역할`이 유추되지 않는다고 생각합니다.
그래서 뷰컨 네이밍에 `역할`이 더 잘 드러날 수 있도록 리팩토링했습니다! 👍🏻 

- NotesSplitViewController -> MemoSplitViewController
- PrimaryTableViewController -> MemoTableViewController
- SecondaryViewController -> MemoDetailViewController

corykim0829 commented on CloudNotes/CloudNotes/Controller/NotesSplitViewController.swift:
```diff
@@ -0,0 +1,39 @@
+        preferredSplitBehavior = .tile
+        preferredDisplayMode = .oneBesideSecondary
+        setViewController(primaryTableViewController, for: .primary)
+        setViewController(secondaryViewController, for: .secondary)
```
UISplitViewController.Column case 때문에 위와 같이 네이밍 하신거면 이 enum 값을 상수화 시키는 것도 좋을 것 같아요 :)

Jager-yoo replied:
`setViewController` 메서드에서 파라미터로 `위치`를 나타내는 primary, secondary 라는 표현이 이미 들어가므로
뷰컨의 인스턴스를 생성할 때, 굳이 그 위치를 다시 표현할 필요는 없다고 생각했습니다!
그래서 아래와 같이 수정했습니다.

제안해주신 enum 값 상수화는 네이밍을 변경하면서 해결된 것 같습니다! 😄 

```swift
// 뷰컨을 인스턴스로 생성해서 변수로 받아주는 부분
private var memoTableViewController = MemoTableViewController(style: .insetGrouped)
private var memoDetailViewController = MemoDetailViewController()

// 스플릿뷰에서 뷰컨의 위치를 지정해주는 부분
setViewController(memoTableViewController, for: .primary)
setViewController(memoDetailViewController, for: .secondary)
```

corykim0829 commented on CloudNotes/CloudNotes/Controller/PrimaryTableViewController.swift:
```diff
@@ -0,0 +1,76 @@
+        let cell = tableView.dequeueReusableCell(withClass: MemoTableViewCell.self, for: indexPath)
+        
+        let data = memo[indexPath.row]
+        cell.configureCellContent(from: data)
```
cell의 뷰를 업데이트 하는 부분 좋습니다 :)

corykim0829 commented on CloudNotes/CloudNotes/Controller/SecondaryViewController.swift:
```diff
@@ -0,0 +1,76 @@
+            return
+        }
+        
+        let userInfo = info as NSDictionary
```
여기서 info를 아예 NSDictionary로 캐스팅하면 더 깔끔할 것 같아요 :)

ChaminLee replied:
말씀주신 것 처럼 타입캐스팅 하는 부분을 한 번에 해주도록 수정해주었습니다!

```swift
guard let userInfo = sender.userInfo as NSDictionary? else {
    return
}
```

corykim0829 commented on CloudNotes/CloudNotes/Controller/SecondaryViewController.swift:
```diff
@@ -0,0 +1,76 @@
+        let keyboardHeight = keyboardFrame.cgRectValue.height
+        
+        memoTextView.contentInset.bottom = keyboardHeight
+        memoTextView.verticalScrollIndicatorInsets.bottom = keyboardHeight
```
깔끔하게 구현 잘하셨습니다 👍

corykim0829 commented on CloudNotes/CloudNotes/Extension/UITableView+dequeueReusableCell.swift:
```diff
@@ -0,0 +1,22 @@
+
+import UIKit
+
+extension UITableView {
```
extension을 활용해서 편리하게 등록한 부분이 좋네요 👍

corykim0829 commented on CloudNotes/CloudNotes/View/MemoTableViewCell.swift:
```diff
@@ -0,0 +1,75 @@
+        self.accessoryType = .disclosureIndicator
+    }
+    
+    func configureCellContent(from memo: Memo) {
```
private 함수들보다는 위에 위치하는게 전체적인 코드를 읽을 때 더 좋을 것 같아요 참고하세요!

ChaminLee replied:
말씀주신 것 처럼 internal 메서드들이 private 메서드들 보다 더 위에 위치하도록 수정하였습니다!

corykim0829 reviewed (changes requested):
수고 많으셨습니다 예거, 차차!! 😆

전반적으로 깔끔하게 잘 구현해서 크게 피드백할 부분이 없었어요 ㅎㅎ 
리뷰 남겨드린 부분 같이 고민해보고 더 개선해봅시다!

---

1️⃣ 지역화 구현 방법

➡️ 시뮬레이터로 진행한 부분인가요? 아이폰으로 실제 기기에서도 Locale 값이 어떻게 나오는지 확인해보면 좋을 것 같아요! 영어로 고정된게 약간의 버그일 것 같은 생각이 듭니다!
저도 개인적으로는 `사용자의 기기 언어를 기준으로 지역화를 하는 게, 지역을 기준으로 하는 것 보다 더 정확하다고 생각하는데요` 이 생각에 동의합니다! 현업에서는 개발적인 관점 외에도 여러 요소로 인해서 지역을 기준으로 지역화를 할 수도 있을 것 같아요. 

2️⃣ PrimaryVC 에서 선택된 cell 데이터를 SecondaryVC 로 보내는 방법

➡️ 지금처럼 PrimaryVC 와 SecondaryVC가 하나의 인스턴스에서 생성되었기 때문에 PrimaryVC에서 상위 객체로 데이터를 전달하고 다시 SecondaryVC에게 넘겨주는게 좋은 구조 같아요!
이 과정에서 delegate 패턴을 사용해도 되고 지금 처럼 splitVC로 호출한 것도 깔끔하고 좋습니다 👍

3️⃣ Secondary View 의 width 가 조정될 때, UITextView 내부의 텍스트가 어색하게 움직이는 이슈

➡️ 흠 저도 이 부분은 좀 더 알아봐야할 것 같은데요! memoTextView의 좌우 패딩 값을 한번 줘보는 건 어떨까요? 
그리고 아이패드 기본 메모앱도 살펴보니까 페이드인, 페이드아웃 효과를 줘서 이런 어색함이 덜보이게 한 것 같아요. 

이 부분에는 너무 시간을 쓰지 않고 나중에 여유가 되면 더 깊게 파보는게 좋을 것 같아요:)

Jager-yoo committed:
refactor: ViewController 네이밍 수정

- 뷰컨 네이밍에 '역할'이 더 드러나도록 수정함
- NotesSplitViewController -> MemoSplitViewController
- PrimaryTableViewController -> MemoTableViewController
- SecondaryViewController -> MemoDetailViewController

ChaminLee committed:
refactor: keyboardWillShow 메서드 내 불필요한 코드 간결화

- internal func가 private func보다 물리적으로 위에 위치하도록 수정

Jager-yoo committed:
refactor: dateFormatter 언어 설정이 디바이스의 선호 언어를 따르도록 수정

- Locale.preferredLanguages 프로퍼티로 디바이스의 언어를 받아와서 dateFormatter 에 할당함
- dateFormatter 상수를 클로저 방식으로 생성하면서 속성 지정까지 해주는 식으로 수정함
- [dateFormatter 를 정의한 타입 extension] Int -> TimeInterval
- TimeInterval 은 Double 타입의 typealias 이므로, Int 로 가져오게 되면, 소수점이 유실되는 상황이 생길 수 있음. 따라서 모델 타입의 lastModified 상수를 TimeInterval 타입으로 변경함

Jager-yoo commented:
@corykim0829 코리!
코드 리뷰 감사합니다! 차차와 리팩토링 완료했습니다.

### ✅ 지역화 이슈 해결
조언해주신 대로, `아이폰` 기기에서도 동일한 이슈가 발생하는지 실험을 해봤는데요, 마찬가지로 아이폰 언어가 "영어", "일본어" 여도 날짜는 `영어`로 나오더라구요!

구글링을 조금 더 하다가 `preferredLanguages` 를 사용하면 된다는 [힌트](https://hyesunzzang.tistory.com/80)를 얻게 되었는데요!
아래와 같은 코드로 리팩토링하니까, `디바이스의 언어`를 따라가더라구요 !!
이런 방식으로 해결했습니다! 👍🏻 

```swift
/*
디바이스의 선호 언어는 배열인데요, 여기서 첫 번째 순서가 == 디바이스의 언어입니다.
first 메서드를 사용하기 때문에 옵셔널이 나오는데요.
기본 언어가 없는 디바이스는 없기 때문에, nil 이 잡히면 '한국어'를 기본값으로 설정해주도록 했습니다!
*/
let devicePreferredLanguage = Locale.preferredLanguages.first ?? "ko"
dateFormatter.locale = Locale(identifier: devicePreferredLanguage)
```

### ✅ UITextView 내부의 텍스트 움직임 이슈

조언해주신 대로, 이 부분은 우선순위가 높지는 않다고 생각합니다!
추후에 여유가 된다면, 꼭 챙겨보겠습니다! 감사합니다. ❤️
